packages/cli/src/ui/components/shared/text-buffer.ts

```diff
@@ -1002,140 +611,502 @@ export function useTextBuffer({
+      initialCursorOffset,
+    );
+    return {
+      lines: lines.length === 0 ? [''] : lines,
+      cursorRow: initialCursorRow,
+      cursorCol: initialCursorCol,
+      preferredCol: null,
+      undoStack: [],
+      redoStack: [],
+      clipboard: null,
+      selectionAnchor: null,
+    };
+  }, [initialText, initialCursorOffset]);
+
+  const [state, dispatch] = useReducer(textBufferReducer, initialState);
+  const { lines, cursorRow, cursorCol, preferredCol, selectionAnchor } = state;
+
+  const text = useMemo(() => lines.join('\n'), [lines]);
+
+  const visualLayout = useMemo(
+    () => calculateVisualLayout(lines, [cursorRow, cursorCol], viewport.width),
+    [lines, cursorRow, cursorCol, viewport.width],
+  );
+
+  const { visualLines, visualCursor } = visualLayout;
+
+  const [visualScrollRow, setVisualScrollRow] = useState<number>(0);
+
+  useEffect(() => {
+    if (onChange) {
+      onChange(text);
+    }
+  }, [text, onChange]);
+
+  // Update visual scroll (vertical)
+  useEffect(() => {
+    const { height } = viewport;
+    let newVisualScrollRow = visualScrollRow;
+
+    if (visualCursor[0] < visualScrollRow) {
+      newVisualScrollRow = visualCursor[0];
+    } else if (visualCursor[0] >= visualScrollRow + height) {
+      newVisualScrollRow = visualCursor[0] - height + 1;
+    }
+    if (newVisualScrollRow !== visualScrollRow) {
+      setVisualScrollRow(newVisualScrollRow);
+    }
+  }, [visualCursor, visualScrollRow, viewport]);
+
+  const insert = useCallback(
+    (ch: string): void => {
+      if (/[\n\r]/.test(ch)) {
+        dispatch({ type: 'insert', payload: ch });
+        return;
+      }
+      dbg('insert', { ch, beforeCursor: [cursorRow, cursorCol] });
+
+      const minLengthToInferAsDragDrop = 3;
+      if (ch.length >= minLengthToInferAsDragDrop) {
+        let potentialPath = ch;
+        if (
+          potentialPath.length > 2 &&
+          potentialPath.startsWith("'") &&
+          potentialPath.endsWith("'")
+        ) {
+          potentialPath = ch.slice(1, -1);
+        }
+
+        potentialPath = potentialPath.trim();
+        if (isValidPath(unescapePath(potentialPath))) {
+          ch = `@${potentialPath}`;
+        }
+      }
+
+      let currentText = '';
+      for (const char of toCodePoints(ch)) {
+        if (char.codePointAt(0) === 127) {
+          if (currentText.length > 0) {
+            dispatch({ type: 'insert', payload: currentText });
+            currentText = '';
+          }
+          dispatch({ type: 'backspace' });
+        } else {
+          currentText += char;
+        }
+      }
+      if (currentText.length > 0) {
+        dispatch({ type: 'insert', payload: currentText });
+      }
     },
-    [
-      visualCursor,
-      visualLines,
-      preferredCol,
-      lines,
-      currentLineLen,
-      visualToLogicalMap,
-      logicalToVisualMap,
-      cursorCol,
-      cursorRow,
-    ],
+    [cursorRow, cursorCol, isValidPath],
```

Comment:
nit: cursorRow and cursorCol are only needed for the dbg logging which isn't crucial. I'd suggest stripping all the dbg logging and seeing which dependencies can be removed. Generally very few of these methods should now need to take dependencies.